pkg/kn/commands/service/service_describe.go

```diff
@@ -41,23 +122,515 @@ func NewServiceDescribeCommand(p *commands.KnParams) *cobra.Command {
 				return err
 			}
 
-			describeService, err := client.GetService(args[0])
+			service, err := client.GetService(serviceName)
 			if err != nil {
 				return err
 			}
 
-			printer, err := serviceDescribePrintFlags.ToPrinter()
+			// Print out machine readable output if requested
+			if machineReadablePrintFlags.OutputFlagSpecified() {
+				printer, err := machineReadablePrintFlags.ToPrinter()
+				if err != nil {
+					return err
+				}
+				return printer.PrintObj(service, cmd.OutOrStdout())
+			}
+
+			printDetails, err = cmd.Flags().GetBool("details")
 			if err != nil {
 				return err
 			}
-			err = printer.PrintObj(describeService, cmd.OutOrStdout())
+
+			revisionDescs, err := getRevisionDescriptions(client, service, printDetails)
+
+			return describe(cmd.OutOrStdout(), service, revisionDescs)
+		},
+	}
+	flags := command.Flags()
+	commands.AddNamespaceFlags(flags, false)
+	flags.BoolP("details", "d", false, "Show all details.")
+	machineReadablePrintFlags.AddFlags(command)
+	return command
+}
+
+// Main action describing the service
+func describe(w io.Writer, service *v1alpha1.Service, revisions []*revisionDesc) error {
+	dw := printers.NewPrefixWriter(w)
+
+	// Service info
+	writeService(dw, service)
+	dw.WriteLine()
+	if err := dw.Flush(); err != nil {
+		return err
+	}
+
+	// Revisions summary info
+	writeRevisions(dw, revisions)
+	dw.WriteLine()
+	if err := dw.Flush(); err != nil {
+		return err
+	}
+
+	// Condition info
+	writeConditions(dw, service)
+	if err := dw.Flush(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Write out main service information. Use colors for major items.
+func writeService(dw printers.PrefixWriter, service *v1alpha1.Service) {
+	dw.WriteColsLn(printers.LEVEL_0, l("Name"), service.Name)
+	dw.WriteColsLn(printers.LEVEL_0, l("Namespace"), service.Namespace)
+	dw.WriteColsLn(printers.LEVEL_0, l("URL"), extractURL(service))
+	if service.Status.Address != nil {
+		url := service.Status.Address.GetURL()
+		dw.WriteColsLn(printers.LEVEL_0, l("Address"), url.String())
+	}
+	writeMapDesc(dw, printers.LEVEL_0, service.Labels, l("Labels"), "")
```

Comment:
Should labels and annotations be printed in columns rather than a comma separated string? IMO, it would improve readability in case of using a large amount of labels and annotations.

Reply:
You get full line-by-line list when usin `--details`. Otherwise its a single line (comma separated), and truncated if too long.

That way I think its a good compromise between a concise initial overview and the full list with all environment variables.

One could argue to even leave out envs, annos and labes when not using `--details` if the comma separated format doesn't make sense.

Reply:
> truncated if too long.

Ah, sorry! I missed the logic of masking too long line.